src/Database/LSMTree/Internal/Entry.hs

```diff
@@ -49,3 +62,68 @@ instance Bifunctor Entry where
       InsertWithBlob v br -> InsertWithBlob v (g br)
       Mupdate v           -> Mupdate v
       Delete              -> Delete
+
+{-------------------------------------------------------------------------------
+  Injections/projections
+-------------------------------------------------------------------------------}
+
+updateToEntryNormal :: Normal.Update v blob -> Entry v blob
+updateToEntryNormal = \case
+    Normal.Insert v Nothing  -> Insert v
+    Normal.Insert v (Just b) -> InsertWithBlob v b
+    Normal.Delete            -> Delete
+
+entryToUpdateNormal :: Entry v blob -> Maybe (Normal.Update v blob)
+entryToUpdateNormal = \case
+    Insert v           -> Just (Normal.Insert v Nothing)
+    InsertWithBlob v b -> Just (Normal.Insert v (Just b))
+    Mupdate _          -> Nothing
+    Delete             -> Just Normal.Delete
+
+updateToEntryMonoidal :: Monoidal.Update v -> Entry v blob
+updateToEntryMonoidal = \case
+    Monoidal.Insert v  -> Insert v
+    Monoidal.Mupsert v -> Mupdate v
+    Monoidal.Delete    -> Delete
+
+entryToUpdateMonoidal :: Entry v blob -> Maybe (Monoidal.Update v)
+entryToUpdateMonoidal = \case
+    Insert v           -> Just (Monoidal.Insert v)
+    InsertWithBlob _ _ -> Nothing
+    Mupdate v          -> Just (Monoidal.Mupsert v)
+    Delete             -> Just Monoidal.Delete
+
+{-------------------------------------------------------------------------------
+  Value resolution/merging
+-------------------------------------------------------------------------------}
+
+-- | Given a value-merge function, combine entries
+combine :: (v -> v -> v) -> Entry v blobref -> Entry v blobref -> Entry v blobref
+combine _ e@Delete            _                       = e
+combine _ e@Insert {}         _                       = e
+combine _ e@InsertWithBlob {} _                       = e
+combine _   (Mupdate u)       Delete                  = Insert u
+combine f   (Mupdate u)       (Insert v)              = Insert (f u v)
+combine f   (Mupdate u)       (InsertWithBlob v blob) = InsertWithBlob (f u v) blob
+combine f   (Mupdate u)       (Mupdate v)             = Mupdate (f u v)
+
+combinesMonoidal :: (v -> v -> v) -> NonEmpty (Entry v blob) -> Entry v blob
+combinesMonoidal f = foldr1 (combine f) -- short-circuit fold
+
+combinesNormal :: NonEmpty (Entry v blob) -> Entry v blob
+combinesNormal = NE.head
+
+-- | Returns 'Nothing' if the combined entries can not be mapped to an
+-- 'Normal.Update'.
+resolveEntriesNormal ::
+     NonEmpty (Entry v blob)
+  -> Maybe (Normal.Update v blob)
+resolveEntriesNormal es = entryToUpdateNormal (combinesNormal es)
+
+-- | Returns 'Nothing' if the combined entries can not be mapped to an
+-- 'Monoidal.Update'.
+resolveEntriesMonoidal ::
+     (v -> v -> v)
+  -> NonEmpty (Entry v blob)
+  -> Maybe (Monoidal.Update v)
+resolveEntriesMonoidal f es = entryToUpdateMonoidal (combinesMonoidal f es)
```

Comment:
It depends on the batching strategy for lookups whether we'll have access to all relevant entries in a single list, so in that case one might have to use `combinesMonoidal` and `combinesNormal` directly

Reply:
> ... combinesNormal has a precondition that there are no monoidal update entries, so it's a little a bit of a footgun, but as you say we might have to expose it from the module. ...

Note that in the monoidal case, updates do not allow blob references either, so both mappings from `Entry` to `Monoidal.Update` and `Normal.Update` are partial. There's not really a way around it since we have a single representation

Reply:
> Theoretically we could first convert to Update and then do value resolution.

True, that is also an option. Combining entries has the advantage that we can skip conversion from `Entry` to `Update` and back when we're merging, and we might skip a few conversions in case we're combining lookup results

Reply:
Yes, that's also something I considered. After inlining, the nested cases in `combineUpdate <$> entryToUpdate e1 <*> entryToUpdate e2` should get optimised away quite nicely I think, but in more complex usages it's not always that easy for the optimiser, so it's most predictable to just work on the entries (as you are doing).